### PR TITLE
📤 Ensure the output folder existed for qnn lib-generator

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes
In the case of model with multi-components, the component name will be append to the output_model_path. That leads the new path does not exist. 
This PR is used to fix the bug in model components where qnn lib-generator's output_model_path does not exist.

Also extract the common part of appending model name into output_model_path, then it can be reused in other model format like qnn-cpp, qnn-so file.

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/model/utils/model_path_utils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+def resolve_model_path(file_or_dir_path: str, model_filename: str) -> str:
+    """Get the model full path.
+
+    The engine provides output paths to Olive passes that do not contain model extension
+    (these paths are generally locations in the cache). This function will convert such
+    paths to absolute file paths and also ensure the parent directories exist.
+    If the input path is already an ONNX file it is simply returned. Examples:
+
+    resolve_onnx_path("c:/foo/bar.onnx", "model.onnx") -> c:/foo/bar.onnx
```
Comment (guotuofeng) [inline olive/model/utils/model_path_utils.py:20]: correct comments?

File: olive/passes/qnn/conversion.py [hunk 2]
```diff
@@ -108,9 +109,7 @@ def _run_for_config(
         elif model.io_config:
             out_nodes = model.io_config.output_names
 
-        output_model_path = Path(output_model_path).resolve()
-        if output_model_path.suffix != ".cpp":
-            output_model_path = output_model_path.with_suffix(".cpp")
+        output_model_path = resolve_model_path(output_model_path, "model.cpp")
```
Comment (guotuofeng) [inline olive/passes/qnn/conversion.py:112]: seems not related to model_path, it like a code path?

File: olive/model/utils/model_path_utils.py (hunk 1 above)
Comment (trajepl) [reply olive/model/utils/model_path_utils.py]: fixed

File: olive/model/utils/path_utils.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    normalize_path_suffix("c:/foo/bar.onnx", "model.onnx") -> c:/foo/bar.onnx
+    normalize_path_suffix("c:/foo/bar.so", "model.so") -> c:/foo/bar.so
+
+    normalize_path_suffix("c:/foo/bar",  "model.onnx") -> c:/foo/bar/model.onnx
+    normalize_path_suffix("c:/foo/bar",  "model.so") -> c:/foo/bar/model.so
+    """
+    suffix = filename_with_suffix.split(".")[-1]
+    if not suffix:
+        raise ValueError(f"Model filename {filename_with_suffix} does not have a valid extension")
```
Comment (guotuofeng) [inline olive/model/utils/path_utils.py:28]: remove Model?

File: olive/model/utils/path_utils.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import logging
+from pathlib import Path
+
+logger = logging.getLogger(__name__)
```
Comment (xiaoyu-work) [inline olive/model/utils/path_utils.py:9]: Are we using logging here?

File: olive/model/utils/path_utils.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    If the input path is already an ONNX file it is simply returned. Examples:
+
+    normalize_path_suffix("c:/foo/bar.onnx", "model.onnx") -> c:/foo/bar.onnx
+    normalize_path_suffix("c:/foo/bar.so", "model.so") -> c:/foo/bar.so
+
+    normalize_path_suffix("c:/foo/bar",  "model.onnx") -> c:/foo/bar/model.onnx
+    normalize_path_suffix("c:/foo/bar",  "model.so") -> c:/foo/bar/model.so
+    """
+    suffix = filename_with_suffix.split(".")[-1]
+    if not suffix:
```
Comment (xiaoyu-work) [inline olive/model/utils/path_utils.py:27]: Applying `split(".")[-1]` to a str without "." will get str itself.

File: olive/model/utils/path_utils.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+def normalize_path_suffix(file_or_dir_path: str, filename_with_suffix: str) -> str:
+    """Get the file full path.
+
+    The engine provides output paths to Olive passes that do not contain file extension
+    (these paths are generally locations in the cache). This function will convert such
+    paths to absolute file paths and also ensure the parent directories exist.
+    If the input path is already an ONNX file it is simply returned. Examples:
+
+    normalize_path_suffix("c:/foo/bar.onnx", "model.onnx") -> c:/foo/bar.onnx
```
Comment (xiaoyu-work) [inline olive/model/utils/path_utils.py:20]: In which case we will have different file name here?

Comment (trajepl) [reply olive/model/utils/path_utils.py:20]: For some of the passes, the output_model_path may contains the model_name which is defined by user.
image

Comment (trajepl) [reply olive/model/utils/path_utils.py:20]: image

File: olive/model/utils/path_utils.py (hunk 4 above)
Comment (trajepl) [reply olive/model/utils/path_utils.py:9]: Oh, sorry. I missed this. Just removed.

**Review by xiaoyu-work — APPROVED** [review]
